### PR TITLE
[WIP] 0.5.0

### PR DESCRIPTION
**Pull request by fqjony**
* Implemented `elasticsearch-cache` library;
* Added `elasticsearch-cache` mocha tests.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -578,6 +578,26 @@ Object.defineProperties( module.exports, {
...
+     *
+     * @param client - override elasticsearch client
+     * @param ttl - override TTL value
+     * @returns {Object}
+     */
+    value: function getElasticSearchCache(client, ttl) {
+      let ElasticSearchCache = require('./elasticsearch-cache');
+      return new ElasticSearchCache({
+        client: client,
+        ttl: ttl
```
Comment (pemaxim) [inline lib/index.js]: @fqjony Wron paramter `ttl`. Should be `default_ttl`

File: lib/index.js [hunk 2]
```diff
@@ -578,6 +578,26 @@ Object.defineProperties( module.exports, {
...
+    /**
+     * Returns ElasticSearch class constructor
+     *
+     * @param client - override elasticsearch client
+     * @param ttl - override TTL value
+     * @returns {Object}
+     */
+    value: function getElasticSearchCache(client, ttl) {
+      let ElasticSearchCache = require('./elasticsearch-cache');
+      return new ElasticSearchCache({
```
Comment (pemaxim) [inline lib/index.js]: @fqjony It overwrites default constructor `options` if `client` and `ttl` even are not provided here.

Comment (pemaxim) [reply lib/index.js]: I would recommend to use `options` object as parameter. And set its default value: `options = {}`.

```
function getElasticSearchCache(options = {})
```

File: lib/elasticsearch-cache.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
+const _ = require('lodash');
+const colors = require('colors');
+const moment = require('moment');
+const client = require('./elasticsearch')();
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony I would recommend to initialize client in the constructor. And only if client is not provided in arguments.

File: lib/elasticsearch-cache.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    return new Promise((resolve, reject) => {
+
+      if (!key) {
+        let err = 'cache.get: Missed key parameter.';
+
+        console.error(err);
+        resolve({
+          ok: false,
+          err: err
+        });
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `return;` has to be provided to end the function here.

File: lib/elasticsearch-cache.js [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+      }
+
+      if (typeof key === 'object') {
+        key = JSON.stringify(key, Object.getOwnPropertyNames(key));
+      }
+
+      let _id = crypto.createHash('sha1').update(key).digest('hex');
+
+      console.log('cache.get: Requesting cache [%s]', colors.green(key));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony Let's use `debug`. Do not execute `console.log`. Especially in third party npm modules. It clutters up logs outputs/

File: lib/elasticsearch-cache.js [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        } else {
+
+          if (_.get(_doc, 'found')) {
+
+            // Retrieving source from entire document
+            let _source = _.get(_doc, '_source');
+
+            // Trying to parse JSON from a body and resolve it in case it has a correct format
+            try {
+              console.log('cache.get: Successfully received cache [%s]', colors.green(key));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony Please, use `debug`

File: lib/elasticsearch-cache.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+              let _body = JSON.parse(_.get(_source, 'body'));
+
+              // Determine if the document is expired - flush it and respond with empty object.
+              if (moment().format() > moment(_.get(_source, 'date')).add(_.get(_source, 'expires', this.options.default_ttl), 's').format()) {
+
+                console.log('cache.get: Document is expired, so returning empty response [%s].', colors.green(key));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `debug`

File: lib/elasticsearch-cache.js [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+              });
+            } catch (err) {
+              console.error('cache.get: ID - [%s] is duplicated in ES database, something weird happens.', colors.red(key));
+              resolve({
+                ok: false,
+                err: err
+              });
+            }
+          } else {
+            console.log('cache.get: Missed cache doc [%s]', colors.green(key));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `debug`

File: lib/elasticsearch-cache.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+   */
+
+  set(key, body, ttl, additionalItems) {
+
+    return new Promise((resolve, reject) => {
+
+      if (!key) {
+        let err = 'cache.set: Missed key parameter.';
+        console.error(err);
+        reject(err);
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `return;`

File: lib/elasticsearch-cache.js [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+        'service': _.get(process.env, 'GIT_NAME', ''),
+        'branch': _.get(process.env, 'GIT_BRANCH', 'production'),
+        'ENV': branch === 'production' ? 'production' : 'development'
+      }, additionalItems);
+
+      if (typeof additionalItems !== 'undefined') {
+        _doc = Object.assign(_doc, _.map(additionalItems, item => item.toString()));
+      }
+
+      console.log('cache.set: Indexing cache [%s]', colors.green(key));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `debug`

File: lib/elasticsearch-cache.js [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+   * @returns {Promise}
+   */
+  flush(key) {
+
+    return new Promise((resolve, reject) => {
+
+      if (!key) {
+        let err = 'cache.set: Missed key parameter.';
+        console.error(err);
+        reject(err);
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `return;`

File: lib/elasticsearch-cache.js [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+      }
+
+      if (typeof key === 'object') {
+        key = JSON.stringify(key, Object.getOwnPropertyNames(key));
+      }
+
+      let _id = crypto.createHash('sha1').update(key).digest('hex');
+
+      console.log('cache.flush: Flushing cache by ES document id [%s]', colors.green(_id));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `debug`

File: lib/elasticsearch-cache.js [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        "index": index,
+        "type": type,
+        "id": _id
+      }, function haveResponse(err, data) {
+
+        if (err) {
+          console.error('cache.flush: Could not remove the cache [%s].', colors.red(_id));
+          reject(err);
+        } else {
+          console.log('cache.flush: Successfully removed cache [%s]', colors.green(_id));
```
Comment (pemaxim) [inline lib/elasticsearch-cache.js]: @fqjony `debug`

File: test/functional/elasticsearch-cache.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+ * Test ElasticSearch Cache.
+ *
+ */
+
+const _ = require('lodash');
+let ElasticsearchCache = require('root-require')('lib/elasticsearch-cache.js');
+const client = require('root-require')('lib/elasticsearch')();
+
+
+let cache = new ElasticsearchCache();
```
Comment (pemaxim) [inline test/functional/elasticsearch-cache.js]: @fqjony You should use `getElasticsearchCache()` wrapper function in tests too.

**Comment by fqjony** [comment]
@pemaxim fixes complete by your feedback. Please re-review it.

**Comment by fqjony** [comment]
@pemaxim I've implemented `buildIdHash` method with the logic of including git branch to the hash. Also, did small refactoring of `elasticsearch-cache` lib. Please review it.

File: readme.md [hunk 15]
```diff
@@ -40,6 +40,7 @@ Module `boxmls-firebase-admin` environments:
 * `sendError( Error, details, callback )`. Sends detailed Error information to email (and Slack. Optional).
 * `scrollResults( query, documentHandler, callback )`. Handler For Scrolling through large dataset with simple callback system
 * `setupElasticsearchMapping( options, callback )`. Creates index and its mapping from file(s).
+* `getElasticsearchCache( client, ttl )`. Returns ElasticSearch cache object
```
Comment (pemaxim) [inline readme.md]: @fqjony `options` instead `client, ttl`

**Pull request merged**